src/schema/slack/functions/_scripts/src/templates/template_function.ts

```diff
@@ -0,0 +1,153 @@
+import { isCustomType } from "../../../../../../types/mod.ts";
+import SchemaTypes from "../../../../../schema_types.ts";
+import SlackSchemaTypes from "../../../../schema_types.ts";
+import { InternalSlackTypes } from "../../../../types/custom/mod.ts";
+import {
+  FunctionParameter,
+  FunctionProperties,
+  FunctionProperty,
+  FunctionRecord,
+} from "../types.ts";
+import {
+  autogeneratedComment,
+  getSlackCallbackId,
+  renderTypeImports,
+} from "./template_utils.ts";
+import { AllowedTypeValue, AllowedTypeValueObject } from "./types.ts";
+
+const typeMap: Record<string, AllowedTypeValueObject> = {
+  SchemaTypes,
+  SlackTypes: SlackSchemaTypes,
+  InternalSlackTypes,
+};
+
+const schemaTypeMap = Object.entries(typeMap).reduce<Record<string, string>>(
+  (acc, [schemaKey, schemaTypes]) => {
+    for (const schemaType in schemaTypes) {
+      const value: AllowedTypeValue = schemaTypes[schemaType];
+      const type: string = isCustomType(value) ? value.id : value;
+      acc[type] = `${schemaKey}.${schemaType}`;
+    }
+    return acc;
+  },
+  {},
+);
+
+const propertyToTypeScript = (
+  property: FunctionProperty,
+): string => {
+  const typescript = [];
+  typescript.push(
+    `type: ${schemaTypeMap[property.type]}`,
+  );
+  if (property.description) {
+    typescript.push(`description: "${property.description}"`);
+  }
+  if (property.title) {
+    typescript.push(`title: "${property.title}"`);
+  }
+  if ("items" in property) {
+    typescript.push(`items: ${propertyToTypeScript(property.items)}`);
+  }
+  if ("properties" in property) {
+    typescript.push(
+      `properties: ${propertiesToTypeScript(property.properties)}`,
+    );
+    if (property.additionalProperties) {
+      typescript.push(`additionalProperties: ${property.additionalProperties}`);
+    }
+    if (property.required) {
+      typescript.push(`required: ${JSON.stringify(property.required)}`);
+    }
```

Comment:
nit: It'd be nice to still set these, regardless if they're on the property
```suggestion
    typescript.push(
      `additionalProperties: ${property.additionalProperties ?? true}`,
    );
    typescript.push(`required: ${JSON.stringify(property.required ?? [])}`);
```

Reply:
I've implemented this but because of type narrowing I had to do this

```js
if ("additionalProperties" in property && property.additionalProperties) {
  typescript.push(`additionalProperties: ${property.additionalProperties}`);
}
```

Reply:
Sorry, I think maybe the commit suggestion feature is causing some confusion here. What we want is that an `ObjectFunctionProperty` will always set it's `additionalProperties` and `required` keys to their defaults, regardless of if they are `undefined` (since we are saying they're optional). This should be totally valid if it's done within the type guard that guarantees it's an object (in this case `if ("properties" in property)`).

```ts
  if ("properties" in property) {
    // Inside the type guard, we know that `property` is `ObjectFunctionProperty`
    // so we can access keys that exist on that type
    typescript.push(
      `properties: ${propertiesToTypeScript(property.properties)}`,
    );
    typescript.push(
      `additionalProperties: ${property.additionalProperties ?? true}`, // if `undefined`, set to `true`
    );
    typescript.push(`required: ${JSON.stringify(property.required ?? [])}`); // if `undefined`, set to an empty array
  }
```

The latest change not only removes the narrowed type, but it also won't set `additionalProperties` or `required` if they're `undefined`.